hail/python/hail/methods/relatedness/mating_simulation.py

```diff
@@ -1,70 +1,162 @@
-                                                                         original=hl.str(ht.__cols[i][ck]),
-                                                                         mother=hl.missing('int32'),
-                                                                         father=hl.missing('int32'))))
-
-    def make_new_generation(prev_generation_tup, idx):
-        prev_size = prev_generation_tup[1]
-        n_new = hl.int32(hl.floor(prev_size * generation_size_multiplier))
-        new_generation = hl.range(n_new).map(
-            lambda i: hl.struct(s=hl.str('generation_') + hl.str(idx + 1) + hl.str('_idx_') + hl.str(i),
-                                original=hl.missing('str'),
-                                mother=hl.rand_int32(0, prev_size),
-                                father=hl.rand_int32(0, prev_size)))
-        return (new_generation, (prev_size + n_new) if keep_founders else n_new)
-
-    ht = ht.annotate_globals(generations=hl.range(n_rounds).scan(lambda prev, idx: make_new_generation(prev, idx),
-                                                                 (ht.generation_0, hl.len(ht.generation_0))))
-
-    def simulate_mating_calls(prev_generation_calls, new_generation):
-        new_samples = new_generation.map(lambda samp: hl.call(prev_generation_calls[samp.mother][hl.rand_int32(0, 2)],
-                                                              prev_generation_calls[samp.father][hl.rand_int32(0, 2)]))
-        if keep_founders:
-            return prev_generation_calls.extend(new_samples)
-        else:
-            return new_samples
+    ns = mt.count_cols()
+
+    # dict of true nonzero relatedness. indeed by tuples of (id1, id2) where a pair is stored with the larger (later) id first.
+    from collections import defaultdict
+    relatedness = defaultdict(dict)
+
+    def get_rel(s1, s2):
+        smaller, larger = (s1, s2) if s1 < s2 else s2, s1
+        return relatedness[larger].get(smaller, 0.0)
+
+    def set_rel(from_s, to_s, value, fwd, generation_start):
+        relatedness[from_s][to_s] = value
+        if to_s >= generation_start:
+            fwd[to_s][from_s] = value
+
+    samples = [(i, f'founder_{i}', None, None) for i in range(ns)]
+    print(f'simulate_random_mating: {len(samples)} founders, {n_rounds} rounds of mating to do')
+    last_generation_start_idx = 0
+    indices = []
+
+    last_gen_fwd = defaultdict(dict)
+
+    if seed is not None:
+        random.seed(seed)
+    for generation in range(n_rounds):
+        last_generation_end = len(samples)
+        mating_generation_size = last_generation_end - last_generation_start_idx
+
+        new_pairs = int(mating_generation_size * pairs_per_generation_multiplier)
+
+        curr_gen_fwd = defaultdict(dict)
+
+        curr_sample_idx = len(samples)
+        parent_to_child = defaultdict(set)
+        for pair in range(new_pairs):
```

Comment:
not for this PR, but I think you've basically built a pedigree simulator. We should pull that out as a separately useful thing.

Reply:
Yeah, curious what other applications might be.